### DISCUSSION TITLE
consider doing GB emulation using radare2

### DISCUSSION THREAD

**Discussion by condret** [visualboyadvance-m/visualboyadvance-m#1157] [Ideas]
I would appreciate it, if you would  discuss about using libr (radare2) for doing the gameboy emulation and disassembly. Well ok, long time ago  I wanted to debug gameboy with using gdbserver from vbam and radare2 after I created disassembler and analysis plugin gameboy into radare2. Turned out, vbam does not do this. So then @radare and I created later somekind of multiarchitecture emulation language, named ESIL. And now radare2 has a fancy IO API which does not require slow memcpy for bankswitches. It also has a binary-parser API, which detects gb-roms correctly, and not just by file-name as you do, afaik.
Extrapoint: it also tells you number of cpu-cycles more precise than the current implementation.

If you use it, u can remove lots of code, that otherwise must be maintained by you.

**Comment by MaskRay** [comment]
I have played a lot of GBA games and a few GB in junior school. It would be nice to see GB emulated in r2.

As a supplement.

> And now radare2 has a fancy IO API which does not require slow memcpy for bankswitches.

siol. The slide will be available at https://github.com/radareorg/r2con-2017/tree/master/talks in a few days.

> It also has a binary-parser API, which detects gb-roms correctly,

https://github.com/radare/radare2/blob/master/libr/bin/p/bin_ningb.c

**Comment by condret** [comment]
and a register api, and gdb-server afaik

**Comment by rkitover** [comment]
We will look into it, we have a lot of problems with GB and GBC emulation also with our current code, memory leaks, crash bugs, glitches, etc..

Thanks for the info!

**Comment by condret** [comment]
well, atm r2 is great for cpu emulation, there are 2 instructions, that cannot be emulated by the plugin atm, bc they need to interact with the outside-world, which is bad for analysis purposes. Anyway, it has tiny functions for redefining every sub-instruction in the emulation-language and adding new ones. Memory-accessed can be hooked and redefined, bc I had projects such as this one in mind, when @radare and I created the language. This is also related to #107, bc radare2 brings its own elf-parser, which is likely to be better than the current one.
Anyway, if you don't use it for CPU emulation, you should consider using it for gb-rom detection and bankswitch emulation, to make your life easier
